io/directory.go

```diff
@@ -294,3 +297,27 @@ func (d *HAMTDirectory) GetNode() (ipld.Node, error) {
 func (d *HAMTDirectory) GetCidBuilder() cid.Builder {
 	return d.shard.CidBuilder()
 }
+
+// UpgradeableDirectory wraps a Directory interface and provides extra logic
+// to upgrade from its BasicDirectory implementation to HAMTDirectory.
+type UpgradeableDirectory struct {
+	Directory
+}
+
+var _ Directory = (*UpgradeableDirectory)(nil)
+
+// AddChild implements the `Directory` interface. We check when adding new entries
+// if we should switch to HAMTDirectory according to global option(s).
+func (d *UpgradeableDirectory) AddChild(ctx context.Context, name string, nd ipld.Node) error {
+	if UseHAMTSharding {
+		if basicDir, ok := d.Directory.(*BasicDirectory); ok {
```

Comment:
nit: check both types and fail if we have some strange third type.

Reply:
ok, will address in upcoming PR, i already have some commits on top of this code

Reply:
(note: for me, "nit" means "I don't really care but I had this random thought while reading this code"; totally optional)